--- a/byterun/caml/misc.h	2018-10-04 06:38:56.000000000 -0700
+++ b/byterun/caml/misc.patched.h	2021-10-18 10:18:47.000000000 -0700
@@ -249,7 +249,7 @@
 #define CAML_SYS_RENAME(old_name,new_name) rename_os(old_name, new_name)
 #define CAML_SYS_CHDIR(dirname) chdir_os(dirname)
 #define CAML_SYS_GETENV(varname) getenv(varname)
-#define CAML_SYS_SYSTEM(command) system_os(command)
+#define CAML_SYS_SYSTEM(command) (-1)
 #define CAML_SYS_READ_DIRECTORY(dirname,tbl) caml_read_directory(dirname,tbl)

 #else
@@ -303,8 +303,7 @@
   CAML_SYS_PRIM_1(CAML_CPLUGINS_CHDIR,chdir_os,dirname)
 #define CAML_SYS_GETENV(varname)                        \
   CAML_SYS_STRING_PRIM_1(CAML_CPLUGINS_GETENV,getenv,varname)
-#define CAML_SYS_SYSTEM(command)                        \
-  CAML_SYS_PRIM_1(CAML_CPLUGINS_SYSTEM,system_os,command)
+#define CAML_SYS_SYSTEM(command) (-1)
 #define CAML_SYS_READ_DIRECTORY(dirname,tbl)                            \
   CAML_SYS_PRIM_2(CAML_CPLUGINS_READ_DIRECTORY,caml_read_directory,     \
                   dirname,tbl)
