--- Makefile.rules.old	2016-05-03 19:21:09.000000000 +0400
+++ Makefile.rules	2016-05-03 19:21:49.000000000 +0400
@@ -35,7 +35,7 @@
 C_OBJECTS = $(C_SOURCE:$($(PROJECT).dir)/%.c=$(BUILDDIR)/$($(PROJECT).dir)/%.o)
 XEN_OBJECTS = $(C_SOURCE:$($(PROJECT).dir)/%.c=$(BUILDDIR)/xen/$($(PROJECT).dir)/%.o)
 
-STUB_LIB = $(if $(C_OBJECTS),$(BUILDDIR)/dll$(PROJECT)_stubs$(EXTDLL))
+STUB_LIB = 
 
 OPAM_PREFIX=$(shell opam config var prefix)
 XEN_LIB = $(if $($(PROJECT).xen),$(BUILDDIR)/dll$(PROJECT)_stubs_xen$(EXTDLL))
@@ -45,7 +45,7 @@
 
 CMO_OPTS = $($(PROJECT).cmo_opts)
 CMX_OPTS = $($(PROJECT).cmx_opts)
-CMA_OPTS = $(if $(C_OBJECTS),-cclib -l$(PROJECT)_stubs -dllib -l$(PROJECT)_stubs)
+CMA_OPTS = $(if $(C_OBJECTS),-cclib -l$(PROJECT)_stubs)
 SUBPROJECT_DEPS = $($(PROJECT).subproject_deps)
 LOCAL_CMXAS = $(SUBPROJECT_DEPS:%=$(BUILDDIR)/%.cmxa)
 CMXA_OPTS = $(if $(C_OBJECTS),-cclib -l$(PROJECT)_stubs)
@@ -58,8 +58,7 @@
 LIB_TARGETS = $(BUILDDIR)/$(PROJECT).cma \
               $(STUB_LIB) \
               $(XEN_LIB) \
-              $(BUILDDIR)/$(PROJECT).cmxa \
-              $(BUILDDIR)/$(PROJECT).cmxs
+              $(BUILDDIR)/$(PROJECT).cmxa
 LIB_TARGET_EXTRAS = $(if $(STUB_LIB),$(BUILDDIR)/lib$(PROJECT)_stubs.a) \
                     $(if $(XEN_LIB),$(BUILDDIR)/lib$(PROJECT)_stubs_xen.a) \
                     $(BUILDDIR)/$(PROJECT).a
